src/byte_sink.rs

```diff
@@ -0,0 +1,76 @@
+use std::io;
+
+use futures::{Async, Poll, Sink, StartSend, AsyncSink};
+
+use {AsyncWrite, AsyncRead};
+
+const BUFFER_SIZE: usize = 8 * 1024;
+
+/// A sink that consumes AsyncRead values, copying their contents to an
+/// AsyncWrite.
+pub struct ByteSink<W, R> {
+    write: W,
+    read: Option<R>,
+    write_cursor: usize,
+    read_cursor: usize,
+    buffer: [u8; BUFFER_SIZE],
+}
+
+/// Construct a sink that consumes AsyncReads, writing their entire contents to
+/// the supplied AsyncWrite.
+pub fn byte_sink<W, R>(write: W) -> ByteSink<W, R> {
+    ByteSink {
+        write: write,
+        read: None,
+        write_cursor: 0,
+        read_cursor: 0,
+        buffer: [0; BUFFER_SIZE],
```

Comment:
This seems pretty large to return as-is, perhaps it could be moved onto the heap?